### PR TITLE
[ISTHMUS] Add substrait to sql support for join and aggregate.

### PR DESCRIPTION
**Pull request by jinfengni**

### CODE REVIEW

File: isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java [hunk 1]
```diff
@@ -42,7 +42,78 @@ public void simpleTestDecimal() throws Exception {
...
+
+  @Test
+  public void simpleTestAgg() throws Exception {
+    test(
+        "select l_partkey, count(l_tax), COUNT(distinct l_discount) from lineitem group by l_partkey");
+  }
+
+  @Test
+  public void simpleTestAggNoGB() throws Exception {
+    test("select count(l_tax), count(distinct l_discount) from lineitem");
```
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:65]: Can we add a test that uses `approx_count_distinct`? Something like this:
```
test("select approx_count_distinct(l_discount) from lineitem");
```

File: isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java [hunk 2]
```diff
@@ -42,7 +42,78 @@ public void simpleTestDecimal() throws Exception {
...
+    // do not verify pojo in round trip test, because of implicit cast(1 as decimal) is removed in
+    // round trip
+    // conversion.
+    test(
+        "select\n"
+            + "  l_returnflag,\n"
+            + "  l_linestatus,\n"
+            + "  sum(l_quantity) as sum_qty,\n"
+            + "  sum(l_extendedprice) as sum_base_price,\n"
+            + "  sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,\n"
```
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:96]: If you use `1.0` instead of `1` or if you add the cast to decimal, does the roundtrip equality succeed?

File: isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java [hunk 3]
```diff
@@ -58,7 +129,10 @@ private void test(String query) throws Exception {
     Rel pojoRel2 =
         SubstraitRelVisitor.convert(RelRoot.of(relnodeRoot, SqlKind.SELECT), EXTENSION_COLLECTION);
```
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:130]: With fix https://github.com/substrait-io/substrait-java/pull/35, does it work to use `ProtoPlanConverter` instead?

File: isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java (hunk 2 above)
Comment (jinfengni) [reply isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:96]: good idea. Using 1.0 will avoid the implicit cast difference in the round trip conversion.

File: isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java (hunk 1 above)
Comment (jinfengni) [reply isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:65]: for  `select approx_count_distinct(l_discount) from lineitem"`,  from sql -> calcite rel, we have approximate/distinct for count() function, but after calcite rel -> substrait rel , the `approximate` flag is lost during `SubstraitRelVisitor`.

Looks like we have to fix the issue calcite rel -> substrait rel.

Comment (JamesRTaylor) [reply isthmus/src/test/java/io/substrait/isthmus/Substrait2SqlTest.java:65]: Does it work going the other direction (from substrait plan -> sql)? You can generate a substrait plan using `count(distinct l_discount)` and then run this code to replace them with `approx_count_distinct`:
```
    ReplaceCountDistinctWithApprox action = new ReplaceCountDistinctWithApprox();
    Plan newPlan = action.modify(oldPlan).orElse(oldPlan);
```
Finally, see if your code to go from substrait plan to sql writes the correct sql with the `approx_count_distinct` call.

**Comment by JamesRTaylor** [comment]
LGTM

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 4]
```diff
@@ -3,19 +3,27 @@
...
+import io.substrait.isthmus.expression.AggregateFunctionConverter;
 import io.substrait.isthmus.expression.ExpressionRexConverter;
 import io.substrait.isthmus.expression.ScalarFunctionConverter;
-import io.substrait.relation.AbstractRelVisitor;
-import io.substrait.relation.Filter;
-import io.substrait.relation.NamedScan;
-import io.substrait.relation.Project;
-import io.substrait.relation.Rel;
+import io.substrait.proto.AggregateFunction;
+import io.substrait.relation.*;
```
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:10]: Minor nit - remove wildcard import. You may want to increase default in IntelliJ which annoyingly defaults to this after three imports from the same package.

File: isthmus/src/test/java/io/substrait/isthmus/SimplePlansTest.java [hunk 5]
```diff
@@ -26,8 +26,7 @@ public void cd() throws IOException, SqlParseException {
     String[] values = asString("tpch/schema.sql").split(";");
     var creates = Arrays.stream(values).filter(t -> !t.trim().isBlank()).toList();
     // creates.forEach(System.out::println);
-    s.execute(
-        "select l_partkey, sum(distinct L_ORDERKEY) from lineitem group by l_partkey ", creates);
+    s.execute("select l_partkey, sum(L_ORDERKEY) from lineitem group by l_partkey ", creates);
```
Comment (JamesRTaylor) [inline isthmus/src/test/java/io/substrait/isthmus/SimplePlansTest.java:29]: Why this change? Revert if not necessary.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 4 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java]: Removed wildcard import. Thanks for the tips in Intellij setting.

File: isthmus/src/test/java/io/substrait/isthmus/SimplePlansTest.java (hunk 5 above)
Comment (jinfengni) [reply isthmus/src/test/java/io/substrait/isthmus/SimplePlansTest.java]: probably changed during debug & testing. 

reverted.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 6]
```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
...
-                expr ->
-                    expr.accept(
-                        new ExpressionRexConverter(
-                            relOptCluster.getTypeFactory(), scalarFunctionConverter)))
+                gr ->
+                    gr.getExpressions().stream()
+                        .map(expr -> expr.accept(expressionRexConverter))
+                        .toList())
             .toList();
+    List<RexNode> groupExprs =
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:150]: It feels like the grouping set dynamics here are wrong. Do you have a test case to verify the correct conversion?

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 7]
```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
...
+        groupExprLists.stream().flatMap(Collection::stream).collect(Collectors.toList());
+    RelBuilder.GroupKey groupKey = relBuilder.groupKey(groupExprs, groupExprLists);
 
-    return relBuilder.push(child).project(rexList).build();
+    List<AggregateCall> aggregateCalls =
+        aggregate.getMeasures().stream().map(m -> fromMeasure(m)).toList();
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:155]: For this kind of thing, using method handle `.map(this::fromMeasure).toList()` is more readable.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 6 above)
Comment (jacques-n) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:150]: Also, how do we map the dynamics of grouping set identifiers between the Calcite and Substrait? (In Substrait we identify the grouping set the result is from by referencing it via ordinal).

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 8]
```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
...
+        aggregateFunctionConverter.getSqlOperatorFromSubstraitFunc(
+            measure.getFunction().declaration().key(), measure.getFunction().outputType());
+    if (!operator.isPresent()) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unable to find binding for call %s", measure.getFunction().declaration().name()));
+    }
+    List<Integer> argIndex = new ArrayList<>();
+    for (RexNode arg : arguments) {
+      if (!(arg instanceof RexInputRef)) {
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:174]: I think we need to fix this. In the case that there are compound expressions, I believe the rewrite to Calcite needs to be to insert a project with the calculation and refer to it from the aggregate function.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 9]
```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
...
+        throw new UnsupportedOperationException(
+            String.format(
+                "aggregation function only support column as argument. % is not support", arg));
+      }
+      argIndex.add(((RexInputRef) arg).getIndex());
+    }
+
+    boolean distinct =
+        measure.getFunction().invocation()
+            == AggregateFunction.AggregationInvocation.AGGREGATION_INVOCATION_DISTINCT;
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:184]: Seems like we should put a check here to ensure invocation is not some unexpected type.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 10]
```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
...
+
+    SqlAggFunction aggFunction;
+    RelDataType returnType =
+        TypeConverter.convert(relOptCluster.getTypeFactory(), measure.getFunction().getType());
+
+    if (operator.get() == SqlStdOperatorTable.COUNT) {
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:190]: This looks like it should be turned into something. I also don't really understand it. It looks like you're saying if a == foo, set b to foo. Why not just set b to foo in the cases a is one of the defined items.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java [hunk 11]
```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
...
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unsupported agg operator % for agg function % ",
+              operator.get(), measure.getFunction().declaration().name()));
+    }
+
+    int filterArg = -1;
+    if (measure.getPreMeasureFilter().isPresent()) {
+      RexNode filter = measure.getPreMeasureFilter().get().accept(expressionRexConverter);
+      if (!(filter instanceof RexInputRef)) {
```
Comment (jacques-n) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:206]: as above, feels like this should be moved to project.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 6 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:150]: `Substrait2SqlTest::simpleTestGroupingSets` has tests covering query with grouping sets.  including `()`. It also has query with `grouping sets` + `group by`.

Substrait uses ordinal, while Calcite uses bitsets.  When converting from substrait to calcite, ordinals are converted into RexInputReference.  The list of RexInputReference in `groupExprs` will converted into indexes in `bitsets`, when we call:

 `RelBuilder.groupKey(Iterable<? extends RexNode> nodes,
      Iterable<? extends Iterable<? extends RexNode>> nodeLists)`

Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:150]: I just added one group sets test, same column appear in the group sets multiple times. It seems the conversion work correctly.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 7 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:155]: changed.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 8 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:174]: Right, when Calcite converts from sql to RelNode, it would insert a project for compound expression.  The conversion of substrait -> calcite should process in a similar way, by adding a project.

However, for now, the substrait plan comes from Calcite RelNode, which has a project for compound expression.  In future, we would add substrait plan utility similar to Calcite's RelBuild/RexBuilder to generate arbitrary substrait plan/expr,  we are required to handle compound expression.  Without such substrait plan utility, it seems not straightforward to write a test with a substrait plan with compound expression in the agg call's argument.    

Shall we open a issue to track this gap and leave it as a TODO?

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 9 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:184]: If invocation is other type,  `distinct` is `false`.   For what reason, we need put such check?

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 10 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:190]: I changed to:
```
   if (operator.get() instanceof SqlAggFunction) {
      aggFunction = (SqlAggFunction) operator.get();
    }
```

`operator` is `SqlOperator`,  a base class.  I could not directly assign it to `aggFunction`.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 11 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:206]: Similar to compound expression for `argument`.  Maybe as a TODO?

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 8 above)
Comment (jacques-n) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:174]: That's fine but we should state in the exception that this just isn't implemented yet (as opposed to being impossible).

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 9 above)
Comment (jacques-n) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:184]: The problem is someone may introduce another branch of the enum in the future. If they do, this path will blindly ignore it. (Basically, the default case of switch).

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 11 above)
Comment (jacques-n) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:203]: As above, let's make sure the error message makes clear that this is a gap in implementation, not capability.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 8 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:174]: Changed the message in the exception to "something is not implemented yet". 

Also, change the exception  from `UnsupportedOperationException`  to `IllegalArgumentException`, when the operator binding is not successful, or the operator is not the expected type.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 9 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:184]: Changed to use switch statement with default cases.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java (hunk 11 above)
Comment (jinfengni) [reply isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java:203]: Changed error message to "something is not implemented yet".

**Comment by jacques-n** [comment]
This looks good to me. Does this look good to you @JamesRTaylor ?

**Comment by JamesRTaylor** [comment]
LGTM too, @jacques-n
